### ISSUE TITLE
[wip] accessibility audit

### ISSUE THREAD

**Issue by annabu** [jupyter/jupyterlab#911]
Notes (some from a quick audit, others from general FYI)... Work in progress.

- Colors should be audited to be AA compliant (they need to be correct color ratio for the given font size - https://www.w3.org/TR/2008/REC-WCAG20-20081211/#visual-audio-contrast-contrast)
- Tabbing should take you to the expected section (and there should be a visual cue i.e. focus state) that lets you know.
- Icons should have labels (even if they are hidden so that screen readers can read it if necessary).
- Colors should not indicate "state" (color blind people will not be able to see red is an error, green is success). Help text and descriptions should be present.

**Comment by ivanov** [comment]
Thanks for getting the ball rolling on this, @annabu - we need all the help we can get not being familiar with this kind of stuff.

**Comment by willingc** [comment]
@annabu Thanks for taking this on :sunny:

**Comment by annabu** [comment]
For the audit, the big take away is that you cannot successfully use the application without a mouse and that not everything is labeled. Main fixes I see right now are:

1. Add aria elements to markup to help screen readers (See: http://heydonworks.com/practical_aria_examples/)
 - buttons and icons should have aria-label (title attributes does not suffice for accessibility)
 - add aria to reference context (see example link)
 - add aria-hidden for content that does not need to be read by screen readers 
2. Make sure everything has a focus state (http://a11yproject.com/posts/never-remove-css-outlines/)
3. Fix tab order for elements (easier once we get the focus state back in, but the success would be taking all primary actions without using a mouse).

Just some consideration moving forward for things that are built:
- Primary text colors are okay but looks like some labels do not pass the color test (particularly labels that are a colored with white text).  To test the colors, try this website: www.aremycoloursaccessible.com
- There does not seem to be any use of heading tags or generally semantic html. Those elements are very useful for users with screen readers and a general complaint with these users when they are not in it (http://www.heydonworks.com/article/responses-to-the-screen-reader-strategy-survey)
- Modals definitely need to be fixed. It won't let you tab or enter to get out of it. Clicking outside of the modal lets you focus on the text area behind it and start typing. (https://www.smashingmagazine.com/2014/09/making-modal-windows-better-for-everyone/)

I recommend that people also download this chrome extension. It lets you audit the page you're in and see what accessibility problems there are on the page: bit.ly/a11y-devtools

**Comment by willingc** [comment]
@annabu Thanks for doing the audit and sharing your tips on tools and resources too. 🌻 

For all, there's also a very good guide re: accessibility at the US Government's 18F site: https://pages.18f.gov/accessibility/

**Comment by jasongrout** [comment]
> Make sure everything has a focus state (http://a11yproject.com/posts/never-remove-css-outlines/)
> Fix tab order for elements (easier once we get the focus state back in, but the success would be taking all primary actions without using a mouse).

+1! I've found this very frustrating that we don't show focus state and that tabIndex settings are kind of weird, especially for dialogs.  I opened https://github.com/jupyter/jupyterlab/issues/1001 to deal with keyboard issues for dialogs.

**Comment by annabu** [comment]
Here's a good read for why you shouldn't ever clear the focus state: https://sarbbottam.github.io/accessibility,/css/2016/09/29/keyboard-accessibility-and-lost-focus/

**Comment by ivanov** [comment]
cross linking https://github.com/jupyter/notebook/issues/1801 -  specifically [this comment](https://github.com/jupyter/notebook/issues/1801#issuecomment-315661799)

**Comment by Neurrone** [comment]
@ivanov will be happy to provide more detail than that comment if it helps.

**Comment by shshkr** [comment]
[https://github.com/shshkr/html-color-contrsast](url)
I created a general tool for calculating contrast ratio from html files. While it still needs some improvement, I thought this tool could be helpful in the initiative to make JupyterLab more accessible.

**Comment by Neurrone** [comment]
Some other useful things:

* Axe, an open source engine that does automated accessibility checks to catch low hanging fruit. It can be run manually on a page by using the Axe extensions for Chrome and Firefox.
* [React-Axe](https://github.com/dequelabs/react-axe), which runs Axe automatically when the DOM changes. This is extremely useful during development.
* The [jsx a11y EsLint plugin](https://github.com/evcohen/eslint-plugin-jsx-a11y) has some overlap, but has some rules of its own not covered by Axe. This provides faster feedback, as it appears in your IDE.

**Comment by eileencho** [comment]
I've also done an accessibility audit on JupyterLab over the summer at NYU with @VickySteeves, linking here the [complete audit results](https://docs.google.com/spreadsheets/d/1RWEtBtXEF5vQf_vJzqtGWnPbT0VZI_RZ_H7RWHmgyGs/edit#gid=0), as well as a [supplement with examples](https://docs.google.com/document/d/1oSHOqmtua_3vdoJddvo4u3na71cIZlpEgP73w-0uXFU/edit).

**Comment by blink1073** [comment]
Thank you @eileencho and @VickySteeves, this is extremely helpful!

**Comment by oigewan** [comment]
Windows has something called "High Contrast" mode (WHCM), which can be quite useful for users with certain kinds of visual impairment.  That is an OS level setting, and will end up having a rather large effect on the CSS in a web page.  While Jupyterlab seems to mostly work OK with WHCM enabled, I did want to link an article which talks about how developers ensure that elements are still useful:  https://sarahmhigley.com/writing/whcm-quick-tips/
